File: src/running_cli/core/merger.py
from datetime import timedelta
from typing import List, Dict, Any, Tuple
from running_cli.core.models import Activity, MergedActivity

def merge_activities(
    garmin_raw: List[Dict[str, Any]], 
    strava_raw: List[Dict[str, Any]]
) -> List[MergedActivity]:
    """Merge Garmin and Strava raw activities, deduplicating overlaps and prioritizing Strava titles."""
    garmin_activities: List[Activity] = []
    for g in garmin_raw:
        try:
            garmin_activities.append(Activity.from_garmin(g))
        except Exception:
            # Silently skip parse errors or log them (for robustness)
            continue

    strava_activities: List[Activity] = []
    for s in strava_raw:
        try:
            strava_activities.append(Activity.from_strava(s))
        except Exception:
            continue

    merged_results: List[MergedActivity] = []
    matched_strava_ids = set()

    for g_act in garmin_activities:
        # Try to find a matching Strava activity
        match_s_act = None
        for s_act in strava_activities:
            if s_act.id in matched_strava_ids:
                continue
            
            # Check matching criteria:
            # 1. Sport is similar or same
            if g_act.sport != s_act.sport:
                # Allow minor mismatched names if times are identical, but generally must match
                if g_act.sport == "Other" or s_act.sport == "Other":
                    pass
                else:
                    continue
            
            # 2. Start time within 10 minutes
            time_diff = abs((g_act.start_time - s_act.start_time).total_seconds())
            if time_diff > 600:  # 10 minutes
                continue
            
            # 3. Distance within 10% or within 500 meters
            dist_diff = abs(g_act.distance_meters - s_act.distance_meters)
            avg_dist = (g_act.distance_meters + s_act.distance_meters) / 2.0
            
            is_dist_match = False
            if avg_dist > 0:
                is_dist_match = (dist_diff / avg_dist) <= 0.10
            else:
                is_dist_match = dist_diff <= 500
            
            if is_dist_match or dist_diff <= 500:
                match_s_act = s_act
                break

        if match_s_act:
            matched_strava_ids.add(match_s_act.id)
            # Create a merged activity
            # Decision: Prefer Strava title over Garmin auto-generated title
            title = match_s_act.title if match_s_act.title else g_act.title
            
            description = match_s_act.description if match_s_act.description else g_act.description
            location_name = g_act.location_name if g_act.location_name else match_s_act.location_name
            latitude = g_act.latitude if g_act.latitude is not None else match_s_act.latitude
            longitude = g_act.longitude if g_act.longitude is not None else match_s_act.longitude
            merged_results.append(
                MergedActivity(
                    date=g_act.start_time.date(),
                    start_time=g_act.start_time.time(),
                    sport=g_act.sport,
                    title=title,
                    distance_meters=g_act.distance_meters,  # Prefer Garmin raw device GPS distance
                    duration_seconds=g_act.duration_seconds,  # Prefer Garmin device elapsed time
                    avg_hr=g_act.avg_hr if g_act.avg_hr is not None else match_s_act.avg_hr,
                    max_hr=g_act.max_hr if g_act.max_hr is not None else match_s_act.max_hr,
                    elevation_gain_meters=g_act.elevation_gain_meters,
                    garmin_id=g_act.id,
                    strava_id=match_s_act.id,
                    description=description,
                    location_name=location_name,
                    latitude=latitude,
                    longitude=longitude,
                    is_race=g_act.is_race or match_s_act.is_race,
                    sources=["garmin", "strava"]
                )
            )
        else:
            # Garmin-only activity
            merged_results.append(
                MergedActivity(
                    date=g_act.start_time.date(),
                    start_time=g_act.start_time.time(),
                    sport=g_act.sport,
                    title=g_act.title,
                    distance_meters=g_act.distance_meters,
                    duration_seconds=g_act.duration_seconds,
                    avg_hr=g_act.avg_hr,
                    max_hr=g_act.max_hr,
                    elevation_gain_meters=g_act.elevation_gain_meters,
                    garmin_id=g_act.id,
                    strava_id=None,
                    description=g_act.description,
                    location_name=g_act.location_name,
                    latitude=g_act.latitude,
                    longitude=g_act.longitude,
                    is_race=g_act.is_race,
                    sources=["garmin"]
                )
            )

    # Process remaining Strava activities (Strava-only)
    for s_act in strava_activities:
        if s_act.id in matched_strava_ids:
            continue
        
        merged_results.append(
            MergedActivity(
                date=s_act.start_time.date(),
                start_time=s_act.start_time.time(),
                sport=s_act.sport,
                title=s_act.title,
                distance_meters=s_act.distance_meters,
                duration_seconds=s_act.duration_seconds,
                avg_hr=s_act.avg_hr,
                max_hr=s_act.max_hr,
                elevation_gain_meters=s_act.elevation_gain_meters,
                garmin_id=None,
                strava_id=s_act.id,
                description=s_act.description,
                location_name=s_act.location_name,
                latitude=s_act.latitude,
                longitude=s_act.longitude,
                is_race=s_act.is_race,
                sources=["strava"]
            )
        )

    # Sort merged activities chronologically
    merged_results.sort(key=lambda x: x.datetime)
    return merged_results

File: pkg/merger/merger.go
package merger

import (
	"math"
	"sort"
	"time"

	"running_cli/pkg/models"
)

// MergeActivities merges Garmin and Strava activities, deduplicating overlaps
func MergeActivities(garmin []*models.Activity, strava []*models.Activity) []*models.MergedActivity {
	var merged []*models.MergedActivity
	matchedStravaIDs := make(map[int64]bool)

	for _, g := range garmin {
		var matchedStrava *models.Activity

		for _, s := range strava {
			if matchedStravaIDs[s.ID] {
				continue
			}

			// 1. Sport must match (allow "Other" fallback)
			if g.Sport != s.Sport {
				if g.Sport != "Other" && s.Sport != "Other" {
					continue
				}
			}

			// 2. Start time difference must be within 10 minutes (600s)
			// Compare start times using local wall-clock times to avoid timezone offset mismatches (e.g. UTC vs local timezone)
			gLocal := time.Date(g.StartTime.Year(), g.StartTime.Month(), g.StartTime.Day(), g.StartTime.Hour(), g.StartTime.Minute(), g.StartTime.Second(), 0, time.UTC)
			sLocal := time.Date(s.StartTime.Year(), s.StartTime.Month(), s.StartTime.Day(), s.StartTime.Hour(), s.StartTime.Minute(), s.StartTime.Second(), 0, time.UTC)
			timeDiff := math.Abs(gLocal.Sub(sLocal).Seconds())
			if timeDiff > 600 {
				continue
			}

			// 3. Distance difference must be within 10% or within 500 meters
			distDiff := math.Abs(g.DistanceMeters - s.DistanceMeters)
			avgDist := (g.DistanceMeters + s.DistanceMeters) / 2.0
			isDistMatch := false

			if avgDist > 0 {
				isDistMatch = (distDiff / avgDist) <= 0.10
			} else {
				isDistMatch = distDiff <= 500
			}

			if isDistMatch || distDiff <= 500 {
				matchedStrava = s
				break
			}
		}

		if matchedStrava != nil {
			matchedStravaIDs[matchedStrava.ID] = true

			// Choose Strava title/description if present
			title := g.Title
			if matchedStrava.Title != "" {
				title = matchedStrava.Title
			}

			description := g.Description
			if matchedStrava.Description != "" {
				description = matchedStrava.Description
			}

			locationName := g.LocationName
			if locationName == "" {
				locationName = matchedStrava.LocationName
			}

			var latitude, longitude *float64
			if g.Latitude != nil {
				latitude = g.Latitude
			} else {
				latitude = matchedStrava.Latitude
			}

			if g.Longitude != nil {
				longitude = g.Longitude
			} else {
				longitude = matchedStrava.Longitude
			}

			var avgHR, maxHR *float64
			if g.AvgHR != nil {
				avgHR = g.AvgHR
			} else {
				avgHR = matchedStrava.AvgHR
			}

			if g.MaxHR != nil {
				maxHR = g.MaxHR
			} else {
				maxHR = matchedStrava.MaxHR
			}

			links := g.Links
			if links == "" {
				links = matchedStrava.Links
			}

			merged = append(merged, &models.MergedActivity{
				Date:                time.Date(g.StartTime.Year(), g.StartTime.Month(), g.StartTime.Day(), 0, 0, 0, 0, time.Local),
				StartTime:           g.StartTime,
				Sport:               g.Sport,
				Title:               title,
				DistanceMeters:      g.DistanceMeters, // Garmin GPS is preferred
				DurationSeconds:     g.DurationSeconds,
				AvgHR:               avgHR,
				MaxHR:               maxHR,
				ElevationGainMeters: g.ElevationGainMeters,
				GarminID:            &g.ID,
				StravaID:            &matchedStrava.ID,
				Description:         description,
				LocationName:        locationName,
				Latitude:            latitude,
				Longitude:           longitude,
				IsRace:              g.IsRace || matchedStrava.IsRace,
				Links:               links,
				Sources:             []string{"garmin", "strava"},
			})
		} else {
			// Garmin only
			merged = append(merged, &models.MergedActivity{
				Date:                time.Date(g.StartTime.Year(), g.StartTime.Month(), g.StartTime.Day(), 0, 0, 0, 0, time.Local),
				StartTime:           g.StartTime,
				Sport:               g.Sport,
				Title:               g.Title,
				DistanceMeters:      g.DistanceMeters,
				DurationSeconds:     g.DurationSeconds,
				AvgHR:               g.AvgHR,
				MaxHR:               g.MaxHR,
				ElevationGainMeters: g.ElevationGainMeters,
				GarminID:            &g.ID,
				Description:         g.Description,
				LocationName:        g.LocationName,
				Latitude:            g.Latitude,
				Longitude:           g.Longitude,
				IsRace:              g.IsRace,
				Links:               g.Links,
				Sources:             []string{"garmin"},
			})
		}
	}

	// Add remaining Strava-only activities (including local manual activities)
	for _, s := range strava {
		if matchedStravaIDs[s.ID] {
			continue
		}

		merged = append(merged, &models.MergedActivity{
			Date:                time.Date(s.StartTime.Year(), s.StartTime.Month(), s.StartTime.Day(), 0, 0, 0, 0, time.Local),
			StartTime:           s.StartTime,
			Sport:               s.Sport,
			Title:               s.Title,
			DistanceMeters:      s.DistanceMeters,
			DurationSeconds:     s.DurationSeconds,
			AvgHR:               s.AvgHR,
			MaxHR:               s.MaxHR,
			ElevationGainMeters: s.ElevationGainMeters,
			StravaID:            &s.ID,
			Description:         s.Description,
			LocationName:        s.LocationName,
			Latitude:            s.Latitude,
			Longitude:           s.Longitude,
			IsRace:              s.IsRace,
			Links:               s.Links,
			Sources:             []string{s.Provider},
		})
	}

	// Sort chronologically
	sort.Slice(merged, func(i, j int) bool {
		return merged[i].StartTime.Before(merged[j].StartTime)
	})

	return merged
}
